File: python/2023/day11.py
from __future__ import annotations
from lib import Point, get_limits
from itertools import combinations


class Universe(list):
    def parse(input: str) -> Universe:
        return Universe(
            Point(x, y)
            for y, line in enumerate(input.splitlines())
            for x, ch in enumerate(line)
            if ch == "#"
        )

    def expanded(self, age: int) -> Universe:
        min_x, max_x = get_limits(map(lambda k: k.x, self))
        min_y, max_y = get_limits(map(lambda k: k.y, self))
        cols = list(filter(lambda x: all(x != other.x for other in self), range(min_x, max_x + 1)))
        rows = list(filter(lambda y: all(y != other.y for other in self), range(min_y, max_y + 1)))

        def _e(p: Point) -> Point:
            x = (age - 1) * len(list(filter(lambda x: x < p.x, cols)))
            y = (age - 1) * len(list(filter(lambda y: y < p.y, rows)))
            return p + Point(x, y)

        return Universe(map(_e, self))

    def solve(self) -> int:
        return sum(map(lambda c: abs(c[0].x - c[1].x) + abs(c[0].y - c[1].y), combinations(self, 2)))


def part1(input: str) -> int:
    return Universe.parse(input).expanded(2).solve()


def part2(input: str) -> int:
    return Universe.parse(input).expanded(1_000_000).solve()


TEST_INPUT = """...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#....."""

PART1_TESTS = [
    (TEST_INPUT, 374),
]

PART2_TESTS = [
    # (TEST_INPUT, 1030),  # 10 times larger
    # (TEST_INPUT, 8410),  # 100 times larger
    (TEST_INPUT, 82000210),  # 1_000_000 times larger
]

File: rust/2023/src/bin/day11.rs
#[derive(Clone, Copy, PartialEq)]
struct Point {
    pub x: usize,
    pub y: usize,
}

impl From<(usize, usize)> for Point {
    fn from(tuple: (usize, usize)) -> Self {
        Self {
            x: tuple.0,
            y: tuple.1,
        }
    }
}

impl Point {
    fn new(x: usize, y: usize) -> Self {
        Self { x, y }
    }
}

struct Solution {
    universe: Vec<Point>,
}

impl Solution {
    fn expanded(&self, age: usize) -> Self {
        let rate = age.saturating_sub(1);

        let x_extent = self
            .universe
            .iter()
            .fold(0, |acc, galaxy| acc.max(galaxy.x));

        let y_extent = self.universe[self.universe.len() - 1].y;

        let empty_cols = (0..=x_extent)
            .into_par_iter()
            .filter(|x| self.universe.iter().all(|g| g.x != *x))
            .collect::<Vec<_>>();

        let empty_rows = (0..=y_extent)
            .into_par_iter()
            .filter(|y| self.universe.iter().all(|g| g.y != *y))
            .collect::<Vec<_>>();

        let universe = self
            .universe
            .par_iter()
            .map(|galaxy| {
                Point::new(
                    galaxy.x + rate * empty_cols.iter().filter(|&&x| x < galaxy.x).count(),
                    galaxy.y + rate * empty_rows.iter().filter(|&&y| y < galaxy.y).count(),
                )
            })
            .collect();

        Self { universe }
    }

    fn solve(&self) -> usize {
        (0..self.universe.len() - 1)
            .into_par_iter()
            .flat_map_iter(|a| {
                (a + 1..self.universe.len()).map(move |b| (self.universe[a], self.universe[b]))
            })
            .map(|(a, b)| a.x.abs_diff(b.x) + a.y.abs_diff(b.y))
            .sum()
    }
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        let universe = input
            .lines()
            .enumerate()
            .flat_map(|(y, line)| {
                line.chars()
                    .enumerate()
                    .filter_map(|(x, c)| {
                        if c == '#' {
                            Some(Point::new(x, y))
                        } else {
                            None
                        }
                    })
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>();

        Ok(Self { universe })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.expanded(2).solve())
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.expanded(1000000).solve())
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT: &str = "...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....";

    #[test]
    fn test_part1() {
        let solution = Solution::new(INPUT).unwrap();
        let answer = solution.expanded(2).solve();
        assert_eq!(answer, 374);
    }

    #[test]
    fn test_part2_1() {
        let solution = Solution::new(INPUT).unwrap();
        let answer = solution.expanded(10).solve();
        assert_eq!(answer, 1030);
    }

    #[test]
    fn test_part2_2() {
        let solution = Solution::new(INPUT).unwrap();
        let answer = solution.expanded(100).solve();
        assert_eq!(answer, 8410);
    }
}
